File: LeetCode/src/graph/CourseSchedule.java
package graph;

import java.util.ArrayList;
import java.util.List;

public class CourseSchedule {
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        List<List<Integer>> graph = new ArrayList<>();

        for (int i = 0; i < numCourses; i++) {
            graph.add(new ArrayList<>());
        }

        for (int[] pre : prerequisites) {
            int preCourse = pre[1];
            int nextCourse = pre[0];
            graph.get(preCourse).add(nextCourse);
        }

        /*
         * 借助一个标志列表 flags，用于判断每个节点 i （课程）的状态：
         * 未被 DFS 访问：i == 0；
         * 已被其他节点启动的 DFS 访问：i == -1；
         * 已被当前节点启动的 DFS 访问：i == 1。
         */
        int[] flags = new int[numCourses];

        for (int i = 0; i < numCourses; i++) {
            boolean result = dfs(graph, flags, i);
            if (!result) {
                return false;
            }
        }

        return true;
    }

    private boolean dfs(List<List<Integer>> graph, int[] flags, int course) {

        if (flags[course] == 1) return false;
        if (flags[course] == -1) return true;

        flags[course] = 1;

        for (Integer pre : graph.get(course)) {
            if (!dfs(graph, flags, pre)) {
                return false;
            }
        }

        flags[course] = -1;

        return true;
    }

    public static void main(String[] args) {
        CourseSchedule solution = new CourseSchedule();
        int numCourses = 4;
        int[][] prerequisites = {{1, 0}, {2, 0}, {3, 1}, {3, 2}};
        boolean result = solution.canFinish(numCourses, prerequisites);// Output: true
        System.out.println("Can finish all courses: " + result);
    }
}

File: LeetCode_Python/graph/CourseSchedule.py
from collections import deque

class CourseSchedule:
    def canFinish(self, numCourses: int, prerequisites: list[list[int]]) -> bool:
        graph = [[] for _ in range(numCourses)]
        indegree = [0] * numCourses

        for pre in prerequisites:
            graph[pre[1]].append(pre[0])
            indegree[pre[0]] += 1
        
        queue = deque()
        for i in range(numCourses):
            if indegree[i] == 0:
                queue.append(i)
        
        course = 0
        while queue:
            node = queue.popleft()
            course += 1

            for nei in graph[node]:
                indegree[nei] -= 1
                if indegree[nei] == 0:
                    queue.append(nei)
        
        return course == numCourses

if __name__ == "__main__":
    solution = CourseSchedule()
    
    # Test Case 1:
    numCourses1 = 2
    prerequisites1 = [[1, 0]]
    print("Test Case 1:")
    print("Input: numCourses =", numCourses1, ", prerequisites =", prerequisites1)
    print("Expected Output: True")
    print("Actual Output:", solution.canFinish(numCourses1, prerequisites1))
    print()
    
    # Test Case 2:
    numCourses2 = 2
    prerequisites2 = [[1, 0], [0, 1]]
    print("Test Case 2:")
    print("Input: numCourses =", numCourses2, ", prerequisites =", prerequisites2)
    print("Expected Output: False")
    print("Actual Output:", solution.canFinish(numCourses2, prerequisites2))
    print()
    
    # Test Case 3:
    numCourses3 = 4
    prerequisites3 = [[1,0],[2,0],[3,1],[3,2]]
    print("Test Case 3:")
    print("Input: numCourses =", numCourses3, ", prerequisites =", prerequisites3)
    print("Expected Output: True")
    print("Actual Output:", solution.canFinish(numCourses3, prerequisites3))
    print()